source/websocket.c

```diff
@@ -0,0 +1,908 @@
+            "id=%p: Failed to send message up channel, error %d (%s).",
+            (void *)websocket,
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        goto error;
+    }
+
+    websocket->thread_data.is_waiting_for_write_completion = true;
+
+    /* If CLOSE frame was written, that's the last data we'll write */
+    if (wrote_close_frame) {
+        s_stop_writing(websocket, AWS_ERROR_HTTP_WEBSOCKET_CLOSE_FRAME_SENT);
+
+        if (websocket->thread_data.is_shutting_down_and_waiting_for_close_frame_to_be_written) {
+            AWS_LOGF_TRACE(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: CLOSE frame sent, finishing handler shutdown sequence.",
+                (void *)websocket);
+
+            s_finish_shutdown(websocket);
+        }
+    }
+
+    return;
+
+error:
+    if (io_msg) {
+        aws_mem_release(io_msg->allocator, io_msg);
+    }
+
+    s_shutdown_due_to_write_err(websocket, aws_last_error());
+}
+
+/* Encoder's outgoing_payload callback invokes current frame's callback */
+static int s_encoder_stream_outgoing_payload(struct aws_byte_buf *out_buf, bool *out_done, void *user_data) {
+    struct aws_websocket *websocket = user_data;
+    assert(aws_channel_thread_is_callers_thread(websocket->channel_slot->channel));
+    assert(websocket->thread_data.current_outgoing_frame);
+
+    struct outgoing_frame *current_frame = websocket->thread_data.current_outgoing_frame;
+    assert(current_frame->def.stream_outgoing_payload);
+
+    enum aws_websocket_outgoing_payload_state payload_state =
+        current_frame->def.stream_outgoing_payload(websocket, out_buf, current_frame->def.user_data);
+
+    *out_done = (payload_state == AWS_WEBSOCKET_OUTGOING_PAYLOAD_DONE);
+
+    /* TODO: handle user closing connection from callback */
+    return AWS_OP_SUCCESS;
+}
+
+static void s_io_message_write_completed(
+    struct aws_channel *channel,
+    struct aws_io_message *message,
+    int err_code,
+    void *user_data) {
+
+    (void)channel;
+    (void)message;
+    struct aws_websocket *websocket = user_data;
+    assert(aws_channel_thread_is_callers_thread(channel));
+
+    if (err_code == AWS_ERROR_SUCCESS) {
+        AWS_LOGF_TRACE(
+            AWS_LS_HTTP_WEBSOCKET, "id=%p: aws_io_message written to socket, sending more data...", (void *)websocket);
+
+        websocket->thread_data.is_waiting_for_write_completion = false;
+        s_try_write_outgoing_frames(websocket);
+    } else {
+        AWS_LOGF_TRACE(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: aws_io_message did not finish writing to socket, error %d (%s).",
+            (void *)websocket,
+            err_code,
+            aws_error_name(err_code));
+
+        s_shutdown_due_to_write_err(websocket, err_code);
+    }
+}
+
+static int s_handler_process_write_message(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    struct aws_io_message *message) {
+
```

Comment:
TODO: for when we actually add mqtt or something on the other side of this.

Reply:
Can't call websockets complete until it's serving MQTT